roachtest: backupFixture/tpcc/warehouses=5000/incrementals=48 failed

roachtest.backupFixture/tpcc/warehouses=5000/incrementals=48 [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAzureBazel/20563533?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAzureBazel/20563533?buildTab=artifacts#/backupFixture/tpcc/warehouses=5000/incrementals=48) on master @ [14ff833383127fa428b913c5281a8a6c862e2a41](https://github.com/cockroachdb/cockroach/commits/14ff833383127fa428b913c5281a8a6c862e2a41):


```
(sql_runner.go:271).Query: error executing 'SELECT job_id, status, error FROM [SHOW JOBS] WHERE job_type = 'BACKUP' AND description ILIKE 'COMPACT BACKUPS%'': dial tcp 20.106.176.74:26257: connect: connection refused
test artifacts and logs in: /artifacts/backupFixture/tpcc/warehouses=5000/incrementals=48/run_1
```

Parameters:
 - <code>arch=amd64</code>
 - <code>cloud=azure</code>
 - <code>coverageBuild=false</code>
 - <code>cpu=8</code>
 - <code>encrypted=true</code>
 - <code>runtimeAssertionsBuild=false</code>
 - <code>ssd=0</code>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

_Grafana is not yet available for azure clusters_</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*backupFixture/tpcc/warehouses=5000/incrementals=48.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

```
panic: Release called on a BufferPool with in-use buffers
	(1) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:791
	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait
	  | 	pkg/util/ctxgroup/ctxgroup.go:151
	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GroupWorkers
	  | 	pkg/util/ctxgroup/ctxgroup.go:211
	  | github.com/cockroachdb/cockroach/pkg/backup.(*restoreDataProcessor).runRestoreWorkers
	  | 	pkg/backup/restore_data_processor.go:397
	  | github.com/cockroachdb/cockroach/pkg/backup.(*restoreDataProcessor).Start.func3
	  | 	pkg/backup/restore_data_processor.go:229
	  | github.com/cockroachdb/cockroach/pkg/backup.(*restoreDataProcessor).Start.Group.GoCtx.func5
	  | 	pkg/util/ctxgroup/ctxgroup.go:200
	  | golang.org/x/sync/errgroup.(*Group).Go.func1
	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
	Wraps: (2) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:791
	  | github.com/cockroachdb/pebble/sstable/block.(*BufferPool).Release
	  | 	external/com_github_cockroachdb_pebble/sstable/block/buffer_pool.go:192
	  | github.com/cockroachdb/pebble.(*externalIterState).Close
	  | 	external/com_github_cockroachdb_pebble/external_iterator.go:123
	  | github.com/cockroachdb/pebble.(*Iterator).Close
	  | 	external/com_github_cockroachdb_pebble/iterator.go:2417
	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).destroy
	  | 	pkg/storage/pebble_iterator.go:996
	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).Close
	  | 	pkg/storage/pebble_iterator.go:332
	  | github.com/cockroachdb/cockroach/pkg/storage.(*ReadAsOfIterator).Close
	  | 	pkg/storage/read_as_of_iterator.go:41
	  | github.com/cockroachdb/cockroach/pkg/backup.(*restoreDataProcessor).openSSTs.(*restoreDataProcessor).openSSTs.func2.func3
	  | 	pkg/backup/restore_data_processor.go:334
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:791
	  | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish
	  | 	pkg/util/tracing/span.go:157
	  | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Tracer
	  | 	pkg/util/tracing/span.go:203
	  | github.com/cockroachdb/cockroach/pkg/util/tracing.ForkSpan
	  | 	pkg/util/tracing/tracer.go:1581
	  | github.com/cockroachdb/cockroach/pkg/cloud.(*loggingTransport).RoundTrip
	  | 	pkg/cloud/cloud_logging_transport.go:54
	  | google.golang.org/api/transport/http.(*parameterTransport).RoundTrip
	  | 	external/org_golang_google_api/transport/http/dial.go:144
	  | go.opencensus.io/plugin/ochttp.(*traceTransport).RoundTrip
	  | 	external/io_opencensus_go/plugin/ochttp/trace.go:84
	  | go.opencensus.io/plugin/ochttp.statsTransport.RoundTrip
	  | 	external/io_opencensus_go/plugin/ochttp/client_stats.go:57
	  | go.opencensus.io/plugin/ochttp.(*Transport).RoundTrip
	  | 	external/io_opencensus_go/plugin/ochttp/client.go:99
	  | golang.org/x/oauth2.(*Transport).RoundTrip
	  | 	external/org_golang_x_oauth2/transport.go:55
	  | net/http.send
	  | 	GOROOT/src/net/http/client.go:259
	  | net/http.(*Client).send
	  | 	GOROOT/src/net/http/client.go:180
	  | net/http.(*Client).do
	  | 	GOROOT/src/net/http/client.go:731
	  | net/http.(*Client).Do
	  | 	GOROOT/src/net/http/client.go:590
	  | cloud.google.com/go/storage.(*httpStorageClient).NewRangeReader.func2.1
	  | 	external/com_google_cloud_go_storage/http_client.go:834
	  | cloud.google.com/go/storage.run.func1
	  | 	external/com_google_cloud_go_storage/invoke.go:67
	  | cloud.google.com/go/internal.retry
	  | 	external/com_google_cloud_go/internal/retry.go:40
	  | cloud.google.com/go/internal.Retry
	  | 	external/com_google_cloud_go/internal/retry.go:33
	  | cloud.google.com/go/storage.run
	  | 	external/com_google_cloud_go_storage/invoke.go:65
	  | cloud.google.com/go/storage.(*httpStorageClient).NewRangeReader.func2
	  | 	external/com_google_cloud_go_storage/http_client.go:833
	  | cloud.google.com/go/storage.(*httpReader).Read
	  | 	external/com_google_cloud_go_storage/http_client.go:1338
	  | cloud.google.com/go/storage.(*Reader).Read
	  | 	external/com_google_cloud_go_storage/reader.go:207
	  | github.com/cockroachdb/cockroach/pkg/cloud.(*ResumingReader).Read
	  | 	pkg/cloud/cloud_io.go:319
	Wraps: (3) assertion failure
	Wraps: (4) attached stack trace
	  -- stack trace:
	  | github.com/cockroachdb/pebble/sstable/block.(*BufferPool).Release
	  | 	external/com_github_cockroachdb_pebble/sstable/block/buffer_pool.go:192
	  | github.com/cockroachdb/pebble.(*externalIterState).Close
	  | 	external/com_github_cockroachdb_pebble/external_iterator.go:123
	  | github.com/cockroachdb/pebble.(*Iterator).Close
	  | 	external/com_github_cockroachdb_pebble/iterator.go:2417
	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).destroy
	  | 	pkg/storage/pebble_iterator.go:996
	  | github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).Close
	  | 	pkg/storage/pebble_iterator.go:332
	  | github.com/cockroachdb/cockroach/pkg/storage.(*ReadAsOfIterator).Close
	  | 	pkg/storage/read_as_of_iterator.go:41
	  | github.com/cockroachdb/cockroach/pkg/backup.(*restoreDataProcessor).openSSTs.(*restoreDataProcessor).openSSTs.func2.func3
	  | 	pkg/backup/restore_data_processor.go:334
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:791
	  | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish
	  | 	pkg/util/tracing/span.go:157
	  | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Tracer
	  | 	pkg/util/tracing/span.go:203
	  | github.com/cockroachdb/cockroach/pkg/util/tracing.ForkSpan
	  | 	pkg/util/tracing/tracer.go:1581
	  | github.com/cockroachdb/cockroach/pkg/cloud.(*loggingTransport).RoundTrip
	  | 	pkg/cloud/cloud_logging_transport.go:54
	  | google.golang.org/api/transport/http.(*parameterTransport).RoundTrip
	  | 	external/org_golang_google_api/transport/http/dial.go:144
	  | go.opencensus.io/plugin/ochttp.(*traceTransport).RoundTrip
	  | 	external/io_opencensus_go/plugin/ochttp/trace.go:84
	  | go.opencensus.io/plugin/ochttp.statsTransport.RoundTrip
	  | 	external/io_opencensus_go/plugin/ochttp/client_stats.go:57
	  | go.opencensus.io/plugin/ochttp.(*Transport).RoundTrip
	  | 	external/io_opencensus_go/plugin/ochttp/client.go:99
	  | golang.org/x/oauth2.(*Transport).RoundTrip
	  | 	external/org_golang_x_oauth2/transport.go:55
	  | net/http.send
	  | 	GOROOT/src/net/http/client.go:259
	  | net/http.(*Client).send
	  | 	GOROOT/src/net/http/client.go:180
	  | net/http.(*Client).do
	  | 	GOROOT/src/net/http/client.go:731
	  | net/http.(*Client).Do
	  | 	GOROOT/src/net/http/client.go:590
	  | cloud.google.com/go/storage.(*httpStorageClient).NewRangeReader.func2.1
	  | 	external/com_google_cloud_go_storage/http_client.go:834
	  | cloud.google.com/go/storage.run.func1
	  | 	external/com_google_cloud_go_storage/invoke.go:67
	  | cloud.google.com/go/internal.retry
	  | 	external/com_google_cloud_go/internal/retry.go:40
	  | cloud.google.com/go/internal.Retry
	  | 	external/com_google_cloud_go/internal/retry.go:33
	  | cloud.google.com/go/storage.run
	  | 	external/com_google_cloud_go_storage/invoke.go:65
	  | cloud.google.com/go/storage.(*httpStorageClient).NewRangeReader.func2
	  | 	external/com_google_cloud_go_storage/http_client.go:833
	  | cloud.google.com/go/storage.(*httpReader).Read
	  | 	external/com_google_cloud_go_storage/http_client.go:1338
	  | cloud.google.com/go/storage.(*Reader).Read
	  | 	external/com_google_cloud_go_storage/reader.go:207
	  | github.com/cockroachdb/cockroach/pkg/cloud.(*ResumingReader).Read
	  | 	pkg/cloud/cloud_io.go:319
	  | github.com/cockroachdb/cockroach/pkg/cloud.(*limitedReader).Read
	  | 	pkg/cloud/impl_registry.go:443
	  | github.com/cockroachdb/cockroach/pkg/ccl/storageccl.(*sstReader).ReadAt
	  | 	pkg/ccl/storageccl/external_sst_reader.go:287
	Wraps: (5) Release called on a BufferPool with in-use buffers
	Error types: (1) *withstack.withStack (2) *withstack.withStack (3) *assert.withAssertionFailure (4) *withstack.withStack (5) *errutil.leafError
```

--- a/pkg/cloud/azure/azure_storage.go
+++ b/pkg/cloud/azure/azure_storage.go
@@ -356,31 +356,47 @@ func isNotFoundErr(err error) bool {
 func (s *azureStorage) ReadFile(
 	ctx context.Context, basename string, opts cloud.ReadOptions,
 ) (_ ioctx.ReadCloserCtx, fileSize int64, _ error) {
+	object := path.Join(s.prefix, basename)
+
 	ctx, sp := tracing.ChildSpan(ctx, "azure.ReadFile")
 	defer sp.Finish()
-	sp.SetTag("path", attribute.StringValue(path.Join(s.prefix, basename)))
-	resp, err := s.getBlob(basename).DownloadStream(ctx, &azblob.DownloadStreamOptions{Range: azblob.
-		HTTPRange{Offset: opts.Offset}})
-	if err != nil {
-		if isNotFoundErr(err) {
-			return nil, 0, cloud.WrapErrFileDoesNotExist(err, "azure blob does not exist")
-		}
-		return nil, 0, errors.Wrapf(err, "failed to create azure reader")
-	}
+	sp.SetTag("path", attribute.StringValue(object))
 
-	if !opts.NoFileSize {
-		if opts.Offset == 0 {
-			fileSize = *resp.ContentLength
-		} else {
-			fileSize, err = cloud.CheckHTTPContentRangeHeader(*resp.ContentRange, opts.Offset)
+	r := cloud.NewResumingReader(ctx,
+		func(ctx context.Context, pos int64) (io.ReadCloser, int64, error) {
+			resp, err := s.getBlob(basename).DownloadStream(ctx, &azblob.DownloadStreamOptions{
+				Range: azblob.HTTPRange{Offset: pos},
+			})
 			if err != nil {
 				return nil, 0, err
 			}
+
+			length := *resp.ContentLength
+			if 0 < pos {
+				length, err = cloud.CheckHTTPContentRangeHeader(*resp.ContentRange, pos)
+				if err != nil {
+					return nil, 0, err
+				}
+			}
+
+			return resp.Body, length, nil
+		},
+		nil, // reader
+		opts.Offset,
+		0,
+		object,
+		cloud.ResumingReaderRetryOnErrFnForSettings(ctx, s.settings),
+		nil, // errFn
+	)
+
+	if err := r.Open(ctx); err != nil {
+		if isNotFoundErr(err) {
+			return nil, 0, cloud.WrapErrFileDoesNotExist(err, "azure blob does not exist")
 		}
+		return nil, 0, err
 	}
-	// BUG: we should follow the azure retry setting here.
-	reader := resp.NewRetryReader(ctx, &azblob.RetryReaderOptions{MaxRetries: 3})
-	return ioctx.ReadCloserAdapter(reader), fileSize, nil
+
+	return r, r.Size, nil
 }
 
 func (s *azureStorage) List(ctx context.Context, prefix, delim string, fn cloud.ListingFn) error {

